packages/runtime/src/useSSRComputation_Client.ts

```diff
@@ -0,0 +1,150 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+import {
+  calculateCacheKey,
+  ClientComputationFunction,
+  Dependency,
+  NoResult,
+  NoResultType,
+  Options,
+  parseDependencies,
+  SSRComputationModule,
+  Subscription,
+} from "./utils";
+import { wrapErrorHandler } from "./errorHandler";
+import { getSSRCache } from "./ssrCache";
+import { runOnSubscriptionsResumed } from "./subscriptions";
+
+type ClientState<TResult> = {
+  module?: SSRComputationModule<TResult>,
+  currentResult: TResult | null,
+  importFn: ClientComputationFunction<TResult>,
+  currentExecutor?: ComputationExecutor<TResult>,
+};
+
+class ComputationExecutor<TResult> {
+  isDisposed = false;
+  subscription: Subscription | undefined;
+
+  constructor(
+    private readonly dependencies: Dependency[],
+    private readonly updateResult: (TResult) => void,
+    private readonly state: ClientState<TResult>,
+  ) {}
+
+  handleSubscriptionIfModuleLoaded = ({ recomputeTheResult }: { recomputeTheResult: boolean }) => {
+    if (this.isDisposed || !this.state.module) return;
+
+    const updateResult = (newResult: TResult | NoResultType) => {
+      if (this.isDisposed || this.state.currentResult === newResult || newResult === NoResult) return;
+      this.updateResult(newResult);
+    }
+
+    if (recomputeTheResult) {
+      const fn = this.state.module?.compute;
+      if (fn) {
+        updateResult(fn(...this.dependencies));
+      }
+    }
+    const getCurrentResult = () => this.state.currentResult;
+    this.subscription = this.state.module?.subscribe?.(getCurrentResult, updateResult, ...this.dependencies);
+  }
+
+  getResultIfModuleLoaded = () => {
+    const fn = this.state.module?.compute;
+    return fn ? fn(...this.dependencies) : NoResult;
+  }
+
+  loadAndRun = () => {
+    if (this.isDisposed) return;
+    this.state.importFn().then(module => {
+      if (this.isDisposed) return;
+      if (!module.compute) throw new Error('The SSR Computation module must have a compute function');
+      this.state.module = module;
+      this.handleSubscriptionIfModuleLoaded({ recomputeTheResult: true });
+    });
+  }
+
+  dispose = () => {
+    if (this.isDisposed) return;
+    this.isDisposed = true;
+    this.subscription?.unsubscribe();
+    this.subscription = undefined;
+  }
+}
+
+const useSSRComputation_Client = <TResult>(
```

Comment:
The component still needs to be simplified. All the magic should happen inside `ClientState`.
The React hook should notify `ClientState` when it's rendered and when it's unmounted and get the return value.

All the magic should be encapsulated inside `ClientState`. The component shouldn't touch no `ClientState's` internals